### PR TITLE
[BugFix] only concern primitive type when comparing ConstantOperator

### LINKED ISSUE #7276: Empty result returned due to predicate range calculation error
ref: https://forum.starrocks.com/t/topic/2659/3

### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #7276 
## Problem Summary(Required) ：

Before calculating predicate range, we may apply `ImplicitCastRule` to cast constant value to specific type. In this case, `name=12345` will be casted to `name=cast(12345 as VARCHAR(1048576))` and `name='12345'` won't be casted, `''12345'`'s type is VARCHAR with default length 65533.

In `ScalarRangePredicateExtractor`, we use `LinkedHashSet<ConstantOperator>` to compute the intersection:
image

In `ConstantOperator::equals` and `ConstantOperator::hashCode`, they're both care about `type`,so the length is considered. In fact, when comparing ConstantOperator, we don't need to care about the length of type, only value and primitive type are enough.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java|2|2|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java [hunk 1]
```diff
@@ -286,7 +286,7 @@ public boolean equals(Object obj) {
         ConstantOperator that = (ConstantOperator) obj;
         return isNull == that.isNull &&
                 Objects.equals(value, that.value) &&
-                Objects.equals(type, that.type);
+                Objects.equals(type.getPrimitiveType(), that.type.getPrimitiveType());
```
[the remark this answers is not in the document]
Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java]: > type.match(type)? PrimitiveType maybe is different, like char, varchar, string....

make sense……I have fixed it

**Review by Youngwb — APPROVED** [review]

**Comment by silverbullet233** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ConstantOperator.java|2|2|100.00%|

**Review by kangkaisen — APPROVED** [review]

**Comment by silverbullet233** [comment]
@Mergifyio backport branch-2.3

**Comment by silverbullet233** [comment]
@Mergifyio backport branch-2.2

**Pull request merged**